paddle/fluid/framework/ir/pass.cc

```diff
@@ -66,30 +68,103 @@ Graph* Pass::Apply(Graph* graph) const {
 
+  ApplyImpl(main_program, startup_program);
+  VLOG(10) << "finish to apply pass " << Type() << " to program";
+}
+
+template <typename Container, typename Visitor>
+static void VisitAllElements(Container &&container, Visitor &&visitor,
+                             bool reverse) {
+  if (reverse) {
+    std::for_each(container.rbegin(), container.rend(), visitor);
+  } else {
+    std::for_each(container.begin(), container.end(), visitor);
+  }
+}
+
+void Pass::MergePrograms(ProgramDesc *dst, const details::ProgramDescs &srcs,
+                         bool append) {
+  PADDLE_ENFORCE_NOT_NULL(
+      dst, platform::errors::InvalidArgument("Dst program must be provided."));
+  bool reverse = !append;
+
+  auto create_var_visitor = [dst](const ProgramDesc &src) {
+    PADDLE_ENFORCE_EQ(src.Size(), 1, platform::errors::Unimplemented(
+                                         "MergePrograms can only support to "
+                                         "merge program with only one block."));
+    const auto &src_block = src.Block(0);
+    auto *dst_block = dst->MutableBlock(0);
+    for (const auto *src_new_var : src_block.AllVars()) {
+      if (dst_block->FindVar(src_new_var->Name())) continue;
+      auto *dst_new_var = dst_block->Var(src_new_var->Name());
+      *dst_new_var = *src_new_var;
+      VLOG(10) << "Create new variable " << dst_new_var->Name();
+    }
+  };
+  VisitAllElements(srcs, create_var_visitor, reverse);
+
+  auto create_op_visitor = [dst, reverse](const ProgramDesc &src) {
+    auto ops = src.Block(0).AllOps();
+    auto copy_op_visitor = [dst, reverse](const OpDesc *src_op) {
+      auto *dst_block = dst->MutableBlock(0);
+      auto *op = reverse ? dst_block->PrependOp() : dst_block->AppendOp();
+      op->CopyFrom(*src_op);
+      VLOG(10) << (reverse ? "Prepend" : "Append") << " op " << op->Type();
+      // FIXME(zjl): some passes does not add VarDesc to program,
+      // we should fix this bug later...
+      for (const auto &in_var_name : op->InputArgumentNames()) {
+        dst_block->Var(in_var_name);
+      }
+      for (const auto &out_var_name : op->OutputArgumentNames()) {
+        dst_block->Var(out_var_name);
+      }
+    };
+    VisitAllElements(ops, copy_op_visitor, reverse);
+  };
+  VisitAllElements(srcs, create_op_visitor, reverse);
+}
+
+void Pass::ApplyImpl(ProgramDesc *main_program,
+                     ProgramDesc *startup_program) const {
   Graph graph(*main_program);
   Apply(&graph);
 
-  // TODO(zjl): support details::kStartupProgramDescs and details::kProgramDescs
   ProgramDesc new_main_program;
   GraphToProgram(graph, &new_main_program);
   main_program->CopyFrom(*new_main_program.Proto());
 
+  if (graph.Has(details::kStartupProgramDescs)) {
+    const auto &startups =
+        graph.Get<details::ProgramDescs>(details::kStartupProgramDescs);
+    VLOG(10) << "Merge startup programs";
+    MergePrograms(startup_program, startups, /*append=*/true);
```

Comment:
Commonly, startup_program would not lose some vars or ops. We only need to merge the created `kStartupProgramDescs` to the startup_program. Some passes may produce `kStartupProgramDescs` (a list of some ProgramDescs) (see [here](https://github.com/PaddlePaddle/Paddle/blob/79e758c61f50a636e1bb7a6134757a66e0b0f869/paddle/fluid/framework/ir/coalesce_grad_tensor_pass.cc#L547)) and run it once when using ParallelExecutor (see [here](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/framework/details/scope_buffered_ssa_graph_executor.cc#L128)). So when we apply passes to ProgramDescs, startup_program should contain these `kStartupProgramDescs`, so that these new created ops can run and only run once.